tests/test_helper/setup_teardown/upgrade.bash

```diff
@@ -38,30 +39,112 @@ setup_file() {
         fi
     done
 
+    # Make sure that microcluster is fully converged before proceeding.
+    # Performing further actions before the microcluster is ready may lead to
+    # unexpectedly long convergence after a microcluster schema upgrade.
+    for container in $TEST_CONTAINERS; do
+        wait_microovn_online "$container" 60
+    done
+
     # detect and export initial MicroOVN snap revision
     container=$(echo "$TEST_CONTAINERS" | awk '{print $1;}' )
     export MICROOVN_SNAP_REV=""
     MICROOVN_SNAP_REV=$(lxc_exec "$container" "snap list | grep microovn | awk '{print \$3;}'")
     assert [ -n "$MICROOVN_SNAP_REV" ]
 
     if [ -n "$UPGRADE_DO_UPGRADE" ]; then
-        echo "# Upgrading MicroOVN from revision $MICROOVN_SNAP_REV" >&3
-        install_microovn "$MICROOVN_SNAP_PATH" $TEST_CONTAINERS
+        assert [ -n "$CENTRAL_CONTAINERS" ]
+        assert [ -n "$CHASSIS_CONTAINERS" ]
+
+        # Export names used locally on chassis containers for use in
+        # teardown_file().
+        export UPGRADE_NS_NAME="upgrade_ns0"
+        export UPGRADE_VIF_NAME="upgrade_vif0"
+
+        # Set up gateway router, workload and background ping on each chassis.
+        for container in $CHASSIS_CONTAINERS; do
+            local ctn_n
+            ctn_n=$(microovn_extract_ctn_n "$container")
+            microovn_add_gw_router "$container"
+            netns_add "$container" "$UPGRADE_NS_NAME"
+            microovn_add_vif "$container" \
+                "$UPGRADE_NS_NAME" "$UPGRADE_VIF_NAME"
+            ping_start "$container" 10.42.${ctn_n}.1 "$UPGRADE_NS_NAME"
+        done
+
+        echo "# Upgrading MicroOVN from revision $MICROOVN_SNAP_REV "\
+             "on central container(s)." >&3
+        install_microovn "$MICROOVN_SNAP_PATH" $CENTRAL_CONTAINERS
```

Comment:
Upgrading central containers in the `file_setup` may be unexpected for the tests, and in fact breaks the "ovsdb schema upgrade" checks in a very subtle way. The schema check relies on collecting schema version before upgrade
https://github.com/canonical/microovn/blob/873f059582140da4c187ba347cb7f231c3aa69f3/tests/test_helper/bats/ovsdb_schema_upgrade.bats#L33-L45

If the snap is already updated by the time the test gets to it, it will always skip the rest of the check. I wonder if it wouldn't be better to make "datapath uptime" checks into it's own test suite.

Reply:
The `file_setup` upgrade is only performed when `UPGRADE_DO_UPGRADE` is set. This is not the case for the ovsdb schema upgrade test. So how would it break it?

Reply:
Oh, I see, my bad. I didn't notice the `UPGRADE_DO_UPGRADE` usage.